### PR TITLE
Update concrete intrinsic naming conventions

### LINKED ISSUE #513: Simulators with missing gates throw NullReferenceException
## Describe the bug
If a simulator in the C# runtime does not implement the interface for a given intrinsic operation / gate, then when the corresponding operation is called from Q#, the simulator throws a `NullReferenceException`.

## To Reproduce

The **experimental** open systems simulator at https://github.com/microsoft/qsharp-runtime/blob/f6593954ad25658f08e44e03e541d39439c4ba3f/src/Simulation/Simulators/OpenSystemsSimulator/OpenSystemsSimulator.cs#L17 implements the `IType1Core` interface, but not the `IQSharpCore` interface, such that the `IGate_Z` interface is not implemented by `Microsoft.Quantum.Experimental.OpenSystemsSimulator`. If `Z` is called from a Q# program, a `NullReferenceException` is thrown.

## Expected behavior
The `IType1Core` interface should allow for the `Z` operation to be implemented by the `ApplyUncontrolledZ__Body` C# method.

## Actual behavior

```
Unhandled exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Quantum.Intrinsic.Z.<get___Body__>b__9_0(Qubit __in__)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Microsoft.Quantum.Simulation.Core.ICallable<I,O>.Apply(I args)      
   at Microsoft.Quantum.Synthesis.ApplyDirectly1C.<get___Body__>b__31_0(ValueTuple`2 __in__)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Microsoft.Quantum.Simulation.Core.ICallable<I,O>.Apply(I args)      
   at Microsoft.Quantum.Synthesis.Apply1C.<get___Body__>b__15_0(ValueTuple`2 __in__)
   at Microsoft.Quantum.Simulation.Core.OperationPartial`3.<get___Body__>b__22_0(P a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Microsoft.Quantum.Simulation.Core.ICallable.Apply(Object args)      
   at Microsoft.Quantum.Canon.ApplyToEach`1.<get___Body__>b__12_0(ValueTuple`2 __in__)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply[GenO](Object args)
   at Microsoft.Quantum.Simulation.Core.GenericCallable.Apply[O](Object args)
   at Microsoft.Quantum.Simulation.Core.GenericCallable.Microsoft.Quantum.Simulation.Core.ICallable.Apply(Object args)
   at Microsoft.Quantum.Characterization.[omitted].<get___Body__>b__30_0(Int64 __in__)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Microsoft.Quantum.Simulation.Core.ICallable<I,O>.Apply(I args)      
   at Microsoft.Quantum.Characterization.[omitted].<get___Body__>b__11_0(ValueTuple`2 __in__)  
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Microsoft.Quantum.Simulation.Core.ICallable<I,O>.Apply(I args)      
   at Microsoft.Quantum.Experimental.[omitted].<get___Body__>b__23_0(ValueTuple`2 __in__) in [omitted]
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Microsoft.Quantum.Simulation.Core.ICallable<I,O>.Apply(I args)      
   at Microsoft.Quantum.Experimental.[omitted].<get___Body__>b__12_0(QVoid __in__) in [omitted]
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply(I a)
   at Microsoft.Quantum.Simulation.Core.Operation`2.Apply[GenO](Object args)
   at Microsoft.Quantum.Simulation.Common.SimulatorBase.Execute[T,I,O](I args)
   at Microsoft.Quantum.Simulation.Common.SimulatorBase.<>c__DisplayClass64_0`3.<Run>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Quantum.EntryPointDriver.Simulation`3.RunSimulator(IEntryPoint`2 entryPoint, ParseResult parseResult, Func`1 createSimulator)
   at Microsoft.Quantum.EntryPointDriver.Simulation`3.Simulate(DriverSettings settings, IEntryPoint`2 entryPoint, ParseResult parseResult, String simulator)
   at Microsoft.Quantum.EntryPointDriver.Driver`3.Simulate(ParseResult parseResult, String simulator)
   at System.CommandLine.Invocation.CommandHandler.GetResultCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()    
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__20_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass15_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass24_0.<<UseVersionOption>b__0>d.MoveNext()--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__21_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__19_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()      
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass13_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

**Additional context**

Looking at `Z.g.cs`, the issue seems to be the use of `as` to cast the operation factory to an `IGate_Z` instance:

```C#
// Z.g.cs
public Z(IOperationFactory m) : base(m)
{
    this.Gate = m as IGate_Z; // this cast fails, but doesn't throw immediately
}
private protected IGate_Z Gate
{
    get;
}
public override Func<Qubit, QVoid> __Body__ => (__in__) =>
{
    var qubit = __in__;
    Gate.Z__Body(qubit); // 💣
    return QVoid.Instance;
}
```

As @swernli pointed out in offline discussion, the issue can be mitigated by modifying `*.csproj` files:

```
<Project Sdk="Microsoft.Quantum.Sdk/0.15.210223170-alpha">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>netcoreapp3.1</TargetFramework>
        <DefaultSimulator>Microsoft.Quantum.Experimental.OpenSystemsSimulator</DefaultSimulator>
        <IncludeQSharpCorePackages>false</IncludeQSharpCorePackages>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="Microsoft.Quantum.EntryPointDriver" Version="0.15.210223170-alpha" />
        <PackageReference Include="Microsoft.Quantum.Type1.Core" Version="0.15.210223170-alpha" />
    </ItemGroup>
</Project>
```

### LINKED ISSUE #536: Simulator throws NullReferenceException when Q# entry point function fails
## Describe the bug

When a Q# entry point function fails, a NullReferenceException is thrown by the simulator.

This is caused by `SimulatorBase.WriteStackTraceToLog` trying to print a stack trace. The stack trace is intended to be updated by the OnOperationStart event handler. Since no operations are executed (only a function), the stack trace is stays null, causing the NullReferenceException.

## To Reproduce

```qsharp
@EntryPoint()
function RunApp() : Unit {
    fail "example";
}
```

## Expected behavior

Only an ExecutionFailException should be thrown.

## Actual behavior

```
Unhandled exception. Microsoft.Quantum.Simulation.Core.ExecutionFailException: example
Unhandled exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Quantum.Simulation.Common.SimulatorBase.WriteStackTraceToLog(Exception exception, IEnumerable`1 callStack)
   at Microsoft.Quantum.Simulation.Common.SimulatorBase.Execute[T,I,O](I args)
   at Microsoft.Quantum.Simulation.Common.SimulatorBase.<>c__DisplayClass64_0`3.<Run>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Quantum.EntryPointDriver.Simulation`3.RunSimulator(IEntryPoint`2 entryPoint, ParseResult parseResult, Func`1 createSimulator)
   at Microsoft.Quantum.EntryPointDriver.Simulation`3.Simulate(DriverSettings settings, IEntryPoint`2 entryPoint, ParseResult parseResult, String simulator)
   at Microsoft.Quantum.EntryPointDriver.Driver`3.Simulate(ParseResult parseResult, String simulator)
   at System.CommandLine.Invocation.CommandHandler.GetResultCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__20_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass15_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass24_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__21_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__19_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass13_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

## System information
- OS: Windows 10

- .NET Core Version: .NET 5.0.103

- IQ# Version: N/A

- Python Version (if applicable): N/A

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by swernli**
This change updates the naming conventions for concrete intrinsics in generated C# and the corresponding interfaces to avoid "gate" terminology. This makes the infrastructure more generic, easier to invoke, and better for use in places besides quantum instruction specific target packages.

Using the `X` gate as an example, the new naming pattern takes what used to be `IGate_X` and instead calls it `IIntrinsicX`, and for the methods `X__Body` and `X__ControlledBody` the intrinsic name is removed, so they become just `Body` and `ControlledBody`.

This also fixes #513.

### CODE REVIEW

File: src/Simulation/CSharpGeneration/SimulationCode.fs [hunk 1]
```diff
@@ -1052,7 +1051,7 @@ module SimulationCode =
             let rec argsToVars = function
                 | QsTupleItem one -> [one.VariableName |> name]
                 | QsTuple many -> many |> Seq.map argsToVars |> List.concat
-            let callExp = (``ident`` "Gate") <.> (``ident`` specCall, argsToVars args)
+            let callExp = ``((`` (``cast`` ("IIntrinsic" + (userDefinedName None op.FullName.Name)) (``ident`` "Impl")) ``))`` <.> (``ident`` specCall, argsToVars args)
```
Comment (swernli) [inline src/Simulation/CSharpGeneration/SimulationCode.fs:1054]: @cgranade With this change, a simulator that tries to invoke a gate that it doesn't implement the interface for will get an invalid cast error instead of a null dereference. Is that sufficient to address #513 ?

File: src/Simulation/Simulators/ToffoliSimulator/Random.cs [hunk 2]
```diff
@@ -1,36 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (ScottCarda-MS) [inline src/Simulation/Simulators/ToffoliSimulator/Random.cs:1]: Why was Random removed, instead of updated like the rest? Did it get moved somewhere else?

File: src/Simulation/Simulators/QCTraceSimulator/QCTraceSimulator.Primitive.Measure.cs [hunk 3]
```diff
@@ -5,10 +5,11 @@ namespace Microsoft.Quantum.Simulation.Simulators.QCTraceSimulators.Implementati
 {
     using System;
     using Microsoft.Quantum.Simulation.Core;
+    using Microsoft.Quantum.Intrinsic.Interfaces;
 
     public partial class QCTraceSimulatorImpl
     {
-        public virtual Result Measure__Body(IQArray<Pauli> paulis, IQArray<Qubit> targets)
```
Comment (ScottCarda-MS) [inline src/Simulation/Simulators/QCTraceSimulator/QCTraceSimulator.Primitive.Measure.cs:11]: nit: Could we keep the access modifiers on the method definitions throughout this PR? I believe without specifying the access, it will default to `private`, but making that explicit goes a long way with readability.

File: src/Simulation/Simulators/QuantumSimulator/ApplyControlledX.cs [hunk 4]
```diff
@@ -2,12 +2,13 @@
 // Licensed under the MIT License.
 
 using Microsoft.Quantum.Simulation.Core;
+using Microsoft.Quantum.Intrinsic.Interfaces;
 
 namespace Microsoft.Quantum.Simulation.Simulators
 {
     public partial class QuantumSimulator
     {
-        public virtual void ApplyControlledX__Body(Qubit control, Qubit target)
```
Comment (ScottCarda-MS) [inline src/Simulation/Simulators/QuantumSimulator/ApplyControlledX.cs:10]: What is the reason for the change in behavior of these from `public virtual` to the implicit `private`?

File: src/Simulation/Simulators/ToffoliSimulator/Random.cs (hunk 2 above)
Comment (swernli) [reply src/Simulation/Simulators/ToffoliSimulator/Random.cs:1]: This was the C# implementation for a deprecated Q# operation. But the deprecated Q#, seen [here](https://github.com/microsoft/qsharp-runtime/blob/8a93f010dca6ab941ee25c1018cf7e1741fb88a0/src/Simulation/QSharpFoundation/Random.qs#L23), was updated to call into `DrawCategorical` which itself is intrinsic and has a separate implementation. So this can safely be removed instead of kept, since it is effectively dead code.

File: src/Simulation/Simulators/QuantumSimulator/ApplyControlledX.cs (hunk 4 above)
Comment (swernli) [reply src/Simulation/Simulators/QuantumSimulator/ApplyControlledX.cs:10]: This is actually an interesting thing that falls out of using interface methods that have the same name. To differentiate `Body` offered by `IIntrinsicApplyControlledX` from `Body` offered by `IIntrinsicApplyControlledZ`, we need to fully qualify the name by saying which interface it comes from. That's why the new signature becomes `void IIntrinsicApplyControlledX.Body`. If I add `public` to this signature, it gives me a compiler error:
```
The modifier 'public' is not valid for this item [Microsoft.Quantum.Simulators] csharp(CS0106)
```
That is because now that the method is fully qualified, it is known to come from a public interface, which means it is public by definition. Including any access modifier, even one that agrees with the defined one, is considered an error.

File: src/Simulation/Simulators/QCTraceSimulator/QCTraceSimulator.Primitive.Measure.cs (hunk 3 above)
Comment (swernli) [reply src/Simulation/Simulators/QCTraceSimulator/QCTraceSimulator.Primitive.Measure.cs:11]: See my other answer, where it turns out that adding access modifiers to fully qualified signatures is not supported.

**Review by ScottCarda-MS — APPROVED** [review]
